packages/next/src/client/components/segment-cache/cache.ts

```diff
@@ -0,0 +1,475 @@
+    return existingEntry
+  }
+  // Create a pending entry and spawn a request for its data.
+  const pendingEntry: PendingRouteCacheEntry = {
+    canonicalUrl: null,
+    status: EntryStatus.Pending,
+    response: null,
+    tree: null,
+    head: null,
+    // If the request takes longer than a minute, a subsequent request should
+    // retry instead of waiting for this one.
+    //
+    // When the response is received, this value will be replaced by a new value
+    // based on the stale time sent from the server.
+    staleAt: now + 60 * 1000,
+    couldBeIntercepted: false,
+  }
+  const key = task.key
+  const requestPromise = fetchRouteOnCacheMiss(pendingEntry, key)
+  trackPrefetchRequestBandwidth(requestPromise)
+  routeCache.set(key.id, pendingEntry)
+  return pendingEntry
+}
+
+/**
+ * Reads the route cache for a matching entry *and* spawns a request if there's
+ * no match. Because this may issue a network request, it should only be called
+ * from within the context of a prefetch task.
+ */
+export function requestSegmentEntryFromCache(
+  now: number,
+  task: PrefetchTask,
+  route: FulfilledRouteCacheEntry,
+  path: string,
+  accessToken: string
+): SegmentCacheEntry {
+  const existingEntry = readSegmentCacheEntry(now, path)
+  if (existingEntry !== null) {
+    return existingEntry
+  }
+  // Create a pending entry and spawn a request for its data.
+  const pendingEntry: PendingSegmentCacheEntry = {
+    status: EntryStatus.Pending,
+    rsc: null,
+    loading: null,
+    staleAt: route.staleAt,
+    promise: null,
+  }
+  trackPrefetchRequestBandwidth(
+    fetchSegmentEntryOnCacheMiss(
+      route,
+      pendingEntry,
+      task.key,
+      path,
+      accessToken
+    )
+  )
+  segmentCache.set(path, pendingEntry)
+  return pendingEntry
+}
+
+function evictRouteCacheEntryFromCache(key: RouteCacheKey): void {
+  routeCache.delete(key.id)
+}
+
+function evictSegmentEntryFromCache(
+  entry: SegmentCacheEntry,
+  key: string
+): void {
+  segmentCache.delete(key)
+  if (entry.status === EntryStatus.Pending && entry.promise !== null) {
+    // There were listeners for this entry. Resolve them with `null` to indicate
+    // that the prefetch failed. It's up to the listener to decide how to handle
+    // this case.
+    // NOTE: We don't currently propagate the reason the prefetch was canceled
+    // but we could by accepting a `reason` argument.
+    entry.promise.resolve(null)
+    entry.promise = null
+  }
+}
+
+function fulfillRouteCacheEntry(
+  entry: PendingRouteCacheEntry,
+  tree: TreePrefetch,
+  head: React.ReactNode,
+  staleAt: number,
+  couldBeIntercepted: boolean,
+  canonicalUrl: string
+) {
+  const fulfilledEntry: FulfilledRouteCacheEntry = entry as any
+  fulfilledEntry.status = EntryStatus.Fulfilled
+  fulfilledEntry.tree = tree
+  fulfilledEntry.head = head
+  fulfilledEntry.staleAt = staleAt
+  fulfilledEntry.couldBeIntercepted = couldBeIntercepted
+  fulfilledEntry.canonicalUrl = canonicalUrl
```

Comment:
Looks like this is missing `fulfilledEntry.response = null`

Reply:
I think I'll just remove this field until we actually need it. Not sure it even makes sense to stash the entire response object instead of the specific data we might need from it, like the size.